server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java

```diff
@@ -324,6 +324,81 @@ public void testReplicaRecovery() throws Exception {
+        ensureGreen(INDEX_NAME);
+
+        logger.info("--> start node C");
+        final String nodeC = internalCluster().startNode();
+        assertFalse(client().admin().cluster().prepareHealth().setWaitForNodes("3").get().isTimedOut());
+
+        // do sync flush to gen sync id
+        assertThat(client().admin().indices().prepareSyncedFlush(INDEX_NAME).get().failedShards(), equalTo(0));
+
+        // hold peer recovery on phase 2 after nodeB down
+        CountDownLatch allowToCompletePhase1Latch = new CountDownLatch(1);
+        MockTransportService transportService = (MockTransportService) internalCluster().getInstance(TransportService.class, nodeA);
+        transportService.addSendBehavior((connection, requestId, action, request, options) -> {
+            if (PeerRecoveryTargetService.Actions.CLEAN_FILES.equals(action)) {
+                try {
+                    allowToCompletePhase1Latch.await();
+                } catch (InterruptedException e) {
+                    throw new AssertionError(e);
+                }
+            }
+            connection.sendRequest(requestId, action, request, options);
+        });
+
+        logger.info("--> restart node B");
+        internalCluster().restartNode(nodeB,
+            new InternalTestCluster.RestartCallback() {
+                @Override
+                public Settings onNodeStopped(String nodeName) throws Exception {
+                    assertBusy(() -> {
+                        // nodeB stopped, peer recovery from nodeA to nodeC, it will be cancelled after nodeB get started.
+                        RecoveryResponse response = client().admin().indices().prepareRecoveries(INDEX_NAME).execute().actionGet();
+
+                        List<RecoveryState> recoveryStates = response.shardRecoveryStates().get(INDEX_NAME);
+                        List<RecoveryState> nodeARecoveryStates = findRecoveriesForTargetNode(nodeA, recoveryStates);
+                        assertThat(nodeARecoveryStates.size(), equalTo(1));
+                        List<RecoveryState> nodeCRecoveryStates = findRecoveriesForTargetNode(nodeC, recoveryStates);
+                        assertThat(nodeCRecoveryStates.size(), equalTo(1));
+
+                        assertRecoveryState(nodeARecoveryStates.get(0), 0, RecoverySource.EmptyStoreRecoverySource.INSTANCE,
+                            true, Stage.DONE, null, nodeA);
+                        validateIndexRecoveryState(nodeARecoveryStates.get(0).getIndex());
+
+                        assertOnGoingRecoveryState(nodeCRecoveryStates.get(0), 0, PeerRecoverySource.INSTANCE,
+                            false, nodeA, nodeC);
+                        validateIndexRecoveryState(nodeCRecoveryStates.get(0).getIndex());
+                    });
+
+                    return super.onNodeStopped(nodeName);
+                }
+            });
+
+        // wait for peer recovering from nodeA to nodeB to be finished
```

Comment:
It took me some time to work out why this works - I suggest this comment explaining it:

```suggestion
        // wait for peer recovery from nodeA to nodeB which is a no-op recovery so it skips the CLEAN_FILES stage and hence is not blocked
```